velox/functions/sparksql/SplitFunctions.cpp

```diff
@@ -16,124 +16,245 @@
 
 #include <utility>
 
+#include "velox/expression/DecodedArgs.h"
 #include "velox/expression/VectorFunction.h"
 #include "velox/expression/VectorWriters.h"
+#include "velox/functions/lib/Re2Functions.h"
+#include "velox/functions/prestosql/Utf8Utils.h"
 
 namespace facebook::velox::functions::sparksql {
 namespace {
 
-/// This class only implements the basic split version in which the pattern is a
-/// single character
-class SplitCharacter final : public exec::VectorFunction {
+// split(string, delimiter[, limit]) -> array(varchar)
+//
+// Splits string on delimiter and returns an array of size at most limit.
+// delimiter is a string representing regular expression.
+// limit is an integer which controls the number of times the regex is applied.
+// By default, limit is -1.
+class Split final : public exec::VectorFunction {
  public:
-  explicit SplitCharacter(const char pattern) : pattern_{pattern} {
-    static constexpr std::string_view kRegexChars = ".$|()[{^?*+\\";
-    VELOX_CHECK(
-        kRegexChars.find(pattern) == std::string::npos,
-        "This version of split supports single-length non-regex patterns");
-  }
+  Split() {}
 
   void apply(
       const SelectivityVector& rows,
       std::vector<VectorPtr>& args,
       const TypePtr& /* outputType */,
       exec::EvalCtx& context,
       VectorPtr& result) const override {
-    exec::LocalDecodedVector input(context, *args[0], rows);
+    // Get the decoded vectors out of arguments.
+    exec::DecodedArgs decodedArgs(rows, args, context);
+    DecodedVector* strings = decodedArgs.at(0);
+    DecodedVector* delims = decodedArgs.at(1);
+    DecodedVector* limits = (args.size() == 2) ? nullptr : decodedArgs.at(2);
 
     BaseVector::ensureWritable(rows, ARRAY(VARCHAR()), context.pool(), result);
     exec::VectorWriter<Array<Varchar>> resultWriter;
     resultWriter.init(*result->as<ArrayVector>());
-
-    rows.applyToSelected([&](vector_size_t row) {
-      resultWriter.setOffset(row);
-      auto& arrayWriter = resultWriter.current();
-
-      const StringView& current = input->valueAt<StringView>(row);
-      const char* pos = current.begin();
-      const char* end = pos + current.size();
-      const char* delim;
-      do {
-        delim = std::find(pos, end, pattern_);
-        arrayWriter.add_item().setNoCopy(StringView(pos, delim - pos));
-        pos = delim + 1; // Skip past delim.
-      } while (delim != end);
-
-      resultWriter.commit();
-    });
+    // Fast path for (flat, const, const).
+    if (strings->isIdentityMapping() and delims->isConstantMapping() and
+        (!limits or limits->isConstantMapping())) {
+      const auto* rawStrings = strings->data<StringView>();
+      const auto delim = delims->valueAt<StringView>(0);
+      int32_t limit = std::numeric_limits<int32_t>::max();
+      if (limits) {
```

Comment:
Can we extract the `limit` computation to a new inline function? I wonder if there is really user case that the `limit` is not constant, all the test in spark the limit is constant.

Reply:
done to extract `limit` to a inline function, and from my point, even `limit` in most cases is `constant`, we still need to keep
non-constant path due to spark don't declare it only support constant `limit`.

Reply:
Before this patch, the second argument `pattern` is constant, I would suggest the new added `limit` also be constant.
CC @rui-mo

Reply:
Spark allows non-constant input for them, so I assume we need to support it. 

> Before this patch, the second argument pattern is constant

This is a limitation mentioned in https://github.com/facebookincubator/velox/issues/4066 which shall be addressed.